internal/service/livedebugging/livedebugging.go

```diff
@@ -0,0 +1,88 @@
+package livedebugging
+
+import "sync"
+
+type ComponentName string
+type ComponentID string
+type CallbackID string
+
+// DebugCallbackManager is used to manage live debugging callbacks.
+type DebugCallbackManager interface {
+	DebugRegistry
+	// AddCallback sets a callback for a given componentID.
+	// The callback is used to send debugging data to live debugging consumers.
+	AddCallback(callbackID CallbackID, componentID ComponentID, callback func(string))
+	// DeleteCallback deletes a callback for a given componentID.
+	DeleteCallback(callbackID CallbackID, componentID ComponentID)
+}
+
+// DebugDataPublisher is used by components to push information to live debugging consumers.
+type DebugDataPublisher interface {
+	DebugRegistry
+	// Publish sends debugging data for a given componentID.
+	Publish(componentID ComponentID, data string)
+	// IsActive returns true when at least one consumer is listening for debugging data for the given componentID.
+	IsActive(componentID ComponentID) bool
+}
+
+// DebugRegistry is used to keep track of the components that supports the live debugging functionality.
+type DebugRegistry interface {
+	// Register a component by name.
+	Register(componentName ComponentName)
+	// IsRegistered returns true if a component has live debugging support.
+	IsRegistered(componentName ComponentName) bool
+}
+
+type liveDebugging struct {
+	loadMut              sync.RWMutex
+	callbacks            map[ComponentID]map[CallbackID]func(string)
+	registeredComponents map[ComponentName]struct{}
+}
+
+var _ DebugCallbackManager = &liveDebugging{}
+var _ DebugDataPublisher = &liveDebugging{}
+
+// NewLiveDebugging creates a new instance of liveDebugging.
+func NewLiveDebugging() *liveDebugging {
+	return &liveDebugging{
+		callbacks:            make(map[ComponentID]map[CallbackID]func(string)),
+		registeredComponents: make(map[ComponentName]struct{}),
+	}
+}
+
+func (s *liveDebugging) Publish(componentID ComponentID, data string) {
+	s.loadMut.RLock()
+	defer s.loadMut.RUnlock()
+	for _, callback := range s.callbacks[componentID] {
+		callback(data)
```

Comment:
I wonder if there'd be an issue and the component has crashed without deleting the callback. Should we detect that and clean up the callback?

Reply:
The components are not responsible for deleting callbacks, it's the job of the DebugCallbackManager which is only passed to the UI service. The callback is only deleted when the user quits the live debugging page of the component. What do you mean by crashed?

Reply:
It's possible that a component stops abruptly for some reason while it's processing data. E.g. maybe it relies on a connection to some service, and it doesn't handle the connection very well so it `Run()` stops all of the sudden. In that case, would the callback unregistration be handled well, or would it leak? If it leaks, maybe we can clean it up the next time the callback is called?

Also, do you think there'd be an issue if a user has the live debugging page open, and then they update their config to exclude that component? What would happen in this case?

Reply:
Because the callback is handled by the API, it won't leak. If the user closes the page, the callback will be removed.

In the case when you update the config to remove the component, the callback will still be there but there won't be a publisher to push to it. As a result, you won't see any data coming to your stream anymore. If you refresh the page you will receive a 404 error because you are trying to access a page that doesn't exist anymore